cameo/strain_design/core.py

```diff
@@ -0,0 +1,319 @@
+    def __init__(self, id, value):
+        super(ReactionKnockinTarget, self).__init__(id, value)
+
+    def apply(self, model, time_machine=None):
+        if time_machine is None:
+            model.add_reaction(self._value)
+        else:
+            time_machine(do=partial(model.add_reaction, self._value),
+                         undo=partial(model.remove_reactions, self._value, delete=False, remove_orphans=True))
+
+    def to_gnomic(self):
+        accession = Target.to_gnomic(self)
+        feature = Feature(accession=accession, type='reaction')
+        return Ins(feature)
+
+
+class GeneModulationTarget(FluxModulationTarget):
+    __gnomic_feature_type__ = "gene"
+
+    def __init__(self, id, value, reference_value):
+        super(GeneModulationTarget, self).__init__(id, value, reference_value)
+
+    def get_target(self, model):
+        return model.genes.get_by_id(self.id)
+
+
+class GeneKnockoutTarget(GeneModulationTarget):
+    def __init__(self, id):
+        super(GeneKnockoutTarget, self).__init__(id, 0, None)
+
+    def apply(self, model, time_machine=None):
+        target = self.get_target(model)
+        target.knock_out(time_machine=time_machine)
+
+
+class ReactionModulationTarget(FluxModulationTarget):
+    __gnomic_feature_type__ = "reaction"
+
+    def __init__(self, id, value, reference_value):
+        super(ReactionModulationTarget, self).__init__(id, value, reference_value)
+
+    def get_target(self, model):
+        return model.reactions.get_by_id(self.id)
+
+
+class ReactionKnockoutTarget(ReactionModulationTarget):
+    def __init__(self, id):
+        super(ReactionKnockoutTarget, self).__init__(id, 0, None)
+
+    def apply(self, model, time_machine=None):
+        target = self.get_target(model)
+        target.knock_out(time_machine=time_machine)
+
+
+class StrainDesignMethod(object):
+    def __init__(self, *args, **kwargs):
+        super(StrainDesignMethod, self).__init__(*args, **kwargs)
+
+    def __call__(self, *args, **kwargs):
+        self.run(*args, **kwargs)
+
+    def run(self, *args, **kwargs):
+        raise NotImplementedError
+
+
+class StrainDesign(object):
+    def __init__(self, targets):
+        self.targets = DictList(targets)
+
+    def __str__(self):
+        return "".join(str(t) for t in self.targets)
+
+    def __repr__(self):
+        return str(self)
+
+    def __iter__(self):
+        for t in self.targets:
+            yield t
+
+    def __len__(self):
+        return len(self.targets)
+
+    def __eq__(self, other):
+        if isinstance(other, StrainDesign):
+            if len(self) != len(other):
+                return False
+            else:
+                return all(self.targets[i] == other.targets[i] for i in range(self))
+        else:
+            return False
+
+    def apply(self, model, time_machine=None):
+        for target in self.targets:
+            target.apply(model, time_machine)
+
+    def __add__(self, other):
```

Comment:
I think this will raise an error if the two strain designs have one or more targets in common. Maybe use the DictList.union instead?